### PR TITLE
HPA: create after the deployment is ready

### PR DESCRIPTION
**Pull request by baum**
Wait for the endpoint deployment to become ready
before creating the controlling HPA at phase4

Issue
====
See https://bugzilla.redhat.com/show_bug.cgi?id=1885524

Test
===

Run in parallel  `watch kubectl get deployment noobaa-endpoint` and  `noobaa install`. Verify that operator is waiting for the deployment to become ready before reconciling the HPA resource. Expected logs sample:
```
➜  kubectl logs -n noobaa noobaa-operator-558d5d767b-qh2nn      -f | grep "endpoint deployment"
time="2021-06-09T09:18:44Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:18:49Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:18:54Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:18:59Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:19:04Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:19:08Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:19:13Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:19:18Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:19:23Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
time="2021-06-09T09:19:28Z" level=warning msg="⏳ Temporary Error: not enough available replicas in endpoint deployment" sys=noobaa/noobaa
```

Make sure there are no HPA errors in the noobaa namespace events using `kubectl get events -n noobaa`

### CODE REVIEW

**Comment by baum** [comment]
Running Metrics server in the development environment
========================================

Download Metrics server YAML
-------------------------------

[Sample Metrics Server YAML](https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml)

Patch
------

Kubelet certificate needs to be signed by cluster Certificate Authority (or disable certificate validation by passing --kubelet-insecure-tls to Metrics Server)

```
--- components.orig.yaml	2021-05-28 18:33:25.000000000 +0300
+++ components.yaml	2021-06-09 11:06:29.000000000 +0300
@@ -131,6 +131,7 @@ spec:
       - args:
         - --cert-dir=/tmp
         - --secure-port=443
+        - --kubelet-insecure-tls
         - --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
         - --kubelet-use-node-status-port
         - --metric-resolution=15s
```

Deploy
-------

```
➜  metrics-server git:(master) ✗ kubectl create -f components.yaml
serviceaccount/metrics-server created
clusterrole.rbac.authorization.k8s.io/system:aggregated-metrics-reader created
clusterrole.rbac.authorization.k8s.io/system:metrics-server created
rolebinding.rbac.authorization.k8s.io/metrics-server-auth-reader created
clusterrolebinding.rbac.authorization.k8s.io/metrics-server:system:auth-delegator created
clusterrolebinding.rbac.authorization.k8s.io/system:metrics-server created
service/metrics-server created
deployment.apps/metrics-server created
apiservice.apiregistration.k8s.io/v1beta1.metrics.k8s.io created
```

Check status
-------------

```
➜  metrics-server git:(master) ✗ kubectl get pod -n kube-system metrics-server-c5f5f4c85-qqbdl
NAME                             READY   STATUS    RESTARTS   AGE
metrics-server-c5f5f4c85-qqbdl   1/1     Running   0          92s
```


Validate
--------

```
➜  metrics-server git:(master) ✗ kubectl top pods -n kube-system
NAME                                     CPU(cores)   MEMORY(bytes)
coredns-f9fd979d6-9j7bq                  9m           9Mi
coredns-f9fd979d6-hkqjz                  11m          9Mi
etcd-docker-desktop                      61m          27Mi
kube-apiserver-docker-desktop            79m          319Mi
kube-controller-manager-docker-desktop   56m          43Mi
kube-proxy-ztwtt                         1m           12Mi
kube-scheduler-docker-desktop            10m          15Mi
metrics-server-c5f5f4c85-qqbdl           11m          17Mi
storage-provisioner                      4m           5Mi
vpnkit-controller                        1m           16Mi
```

**Review by liranmauda — COMMENTED** [review]
- Let's change the beautify from default to beautify 
- Regarding upgrade, this PR will fix the BZ in a new installation, but it might happen when upgrading as the HPA will already be there and the endpoint are being replaced.

**Comment by nimrod-becker** [comment]
@liranmauda might be good enough to do install only

**Review by liranmauda — APPROVED** [review]

**Comment by guymguym** [comment]
@baum Can you explain why this is needed to wait?
Usually in kubernetes it should be possible to create all CR's in parallel, and the controllers should reconcile eventually.
What is the problem with HPA created when the deployment is not ready?

**Comment by nimrod-becker** [comment]
HPA tried to query the CPU of the endpoints before they were ready, leading to alerts

**Comment by nimrod-becker** [comment]
closed by mistake, sorry

**Comment by guymguym** [comment]
OK, so this is a generic HPA thing - is it well know?
Is there any issue on this in kubernetes?

**Comment by nimrod-becker** [comment]
Might be related to readiness checks for the endpoint pods?
see [here](https://docs.openshift.com/container-platform/3.7/dev_guide/pod_autoscaling.html#hpa-autoscaling)

**Comment by baum** [comment]
@nimrod-becker thank you for the pointer. Tried defining `initialDelaySeconds` as 1 and 5 mins in the deployment-endpoint's `readinessProbe`. I am using noobaa-operator's master version as the base for this test.

The change:

```diff
diff --git a/deploy/internal/deployment-endpoint.yaml b/deploy/internal/deployment-endpoint.yaml 
index 1a8c118..a8d2cfe 100644
--- a/deploy/internal/deployment-endpoint.yaml
+++ b/deploy/internal/deployment-endpoint.yaml
@@ -106,6 +106,7 @@ spec:
             tcpSocket:
               port: 6001 # ready when s3 port is open
             timeoutSeconds: 5
+            initialDelaySeconds: 300
       securityContext:
         runAsUser: 0
         runAsGroup: 0
```

I still see events in the `noobaa-endpoint` HPA:

```
➜  k describe  horizontalpodautoscalers.autoscaling noobaa-endpoint
Name:                                                  noobaa-endpoint
Namespace:                                             default
Labels:                                                app=noobaa
...
Events:
  Type     Reason                        Age                 From                       Message
  ----     ------                        ----                ----                       -------
  Warning  FailedGetResourceMetric       60s (x5 over 5m1s)  horizontal-pod-autoscaler  did not receive metrics for any ready pods
  Warning  FailedComputeMetricsReplicas  60s (x5 over 5m1s)  horizontal-pod-autoscaler  invalid metrics (1 invalid out of 1), first error is: failed to get cpu utilization: did not receive metrics for any ready pods
```
WDYT?

**Comment by nimrod-becker** [comment]
@baum sadly seems like its not the solution we were looking for. @guymguym I didn't find any other mentions regarding this phenomena, most examples and blogs I found were about adding the HPA after the pod(s) were already up. The Readiness Checks were the only hint O saw towards what's happening when the pods are not up yet and their CPU can't be queried

**Comment by baum** [comment]
@nimrod-becker @guymguym  The error is that metrics for the pods are not available. I assume that it could be either if (a) metrics server is not ready or (b) pods are not running. QE negated option (a) that is why option (b) is pursued.  And [agree](https://github.com/noobaa/noobaa-operator/pull/649#issuecomment-872112794) there is an implicit assumption (see [here](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#create-horizontal-pod-autoscaler)) that HPA is created on top of existing Deployment instance.

File: pkg/system/phase4_configuring.go [hunk 1]
```diff
@@ -431,25 +432,52 @@ func (r *Reconciler) setDesiredEndpointMounts(podSpec *corev1.PodSpec, container
...
+	}
+
+	return nil
+}
+
+// ReconcileHPAEndpoint reconcile the endpoint's HPA and report the configuration
 // back to the noobaa core
 func (r *Reconciler) ReconcileHPAEndpoint() error {
+	// Wait for the the endpoint deployment to become ready
+	if err := r.awaitEndpointDeploymentPods(); err != nil {
```
Comment (dannyzaken) [inline pkg/system/phase4_configuring.go:477]: I think that always waiting for endpoints deployment before reconciling HPA, can cause us to get stuck here if the deployment can't reach the desired number of replicas (e.g. if the user sets a high number of min count in the spec).

Comment (baum) [reply pkg/system/phase4_configuring.go]: @dannyzaken  good point, condition to wait only if HPA was not created yet was added.

**Review by dannyzaken — APPROVED** [review]

**Pull request merged**